### ISSUE TITLE
Windows forms ComboBox.DropDownStyle change causes System.AccessViolationException randomly

### ISSUE THREAD

**Issue by Marie-Shi** [dotnet/winforms#3595]
* .NET Core Version: 
.Net 5.0.100-rc.1.20366.7 from 5.0 master branch + VS2019 with latest core designer

* Have you experienced this same bug with .NET Framework?: 
Yes

**More Info**
This issue can reproduce from .NET 3.5 to .NET 4.8 on Dev14/Dev15/Dev16. And this is a randomly issue, not happened each time.

**Minimal repro:**
1.	Extract “Core.zip” and open it in VS.
[Core.zip](https://github.com/dotnet/winforms/files/4935973/Core.zip)
2.	Build and run the project.
3.	Click the dropdown button of ComboBox in form1, then form2 shown.
4.	Click “X” to close form2.

 **Actual behavior:**
After step4, an error dialog pop up: 
[Call_stack.txt](https://github.com/dotnet/winforms/files/4935975/Call_stack.txt)
![System_AccessViolationException](https://user-images.githubusercontent.com/45864017/87753381-1044fd00-c835-11ea-957b-45225cf06de7.png)

**Expected behavior:**
There is no error dialog pop up. The project can run normally.

**Comment by weltkante** [comment]
You can reproduce by a ComboBox which changes its style inside its DropDown callback. No need to open a separate form.

```csharp
private void CBDropDown(object sender, EventArgs e)
{
    comboBox1.DropDownStyle = ComboBoxStyle.Simple;
    comboBox1.DropDownStyle = ComboBoxStyle.DropDown;
}
```

The crash is probably caused because changing style will call `RecreateHandle`. Destroying the control while you are in the middle of its callback I suspect will corrupt memory, because once you return from the callback the native part of the ComboBox will continue executing on an already destroyed native control.

/cc @JeremyKuhne

**Comment by weltkante** [comment]
I wonder if WinForms should introduce a nesting counter per control so it can know if it currently is within a WndProc. Could then either throw an exception or employ smarter control handle recycling (e.g. making the old one invisible + deferred delete instead of deleting it immediately if you are in the WndProc).

**Comment by JeremyKuhne** [comment]
> I wonder if WinForms should introduce a nesting counter per control so it can know if it currently is within a WndProc. Could then either throw an exception or employ smarter control handle recycling (e.g. making the old one invisible + deferred delete instead of deleting it immediately if you are inside its WndProc).

That seems reasonable. Perhaps on `NativeWindow`?

**Comment by RussKie** [comment]
@weltkante is this something you could fix perhaps? 😉

**Comment by weltkante** [comment]
I can start making a draft PR but it probably needs further discussion off of that, above I just outlined the idea how to fix it but to discover the actual implications it has needs prototyping.

**Comment by weltkante** [comment]
Hmm, having trouble getting master to build, it can't find the .NET 6 packages; am I missing something or is this a known issue and I just wait until its sorted out?

<details><summary>build log</summary>

```
D:\projects\winforms2 [master ≡]> .\build
Processing D:\projects\winforms2\global.json
cmake.exe already exists; replacing...
dotnet-api-docs_net5.0 was not found in .
Native tools bootstrap failed
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100/dotnet-sdk-5.0.100-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100/dotnet-sdk-5.0.100-win-x64.zip
dotnet-install: Adding to current process PATH: "D:\projects\winforms2\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install/windows#dependencies
dotnet-install: Installation finished
D:\projects\winforms2\.dotnet\sdk\5.0.100\MSBuild.dll /nologo -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,D:\projects\winforms2\.dotnet\sdk\5.0.100\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,D:\projects\winforms2\.dotnet\sdk\5.0.100\dotnet.dll -maxcpucount /m -verbosity:m /v:minimal /bl:D:\projects\winforms2\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:True /p:ContinuousIntegrationBuild=False /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=D:\projects\winforms2\artifacts\toolset\6.0.0-beta.21078.12.txt /t:__WriteToolsetLocation /warnaserror D:\projects\winforms2\artifacts\toolset\restore.proj
D:\projects\winforms2\artifacts\toolset\restore.proj : error : Das Paket "Microsoft.DotNet.Arcade.Sdk" der Version (= 6.0.0-beta.21078.12) wurde nicht gefunden.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 178 Version(en) gefunden in "dotnet-tools" [ Nächste Version: 5.0.0-beta.19558.13 ]
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 2 Version(en) gefunden in "dotnet5-transport" [ Nächste Version: 5.0.0-beta.20364.3 ]
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in nuget.org.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in richnav.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in dotnet5.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in dotnet6-transport.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in dotnet6.
D:\projects\winforms2\artifacts\toolset\restore.proj : error MSB4236: Das angegebene SDK "Microsoft.DotNet.Arcade.Sdk" wurde nicht gefunden.
Build failed with exit code 1. Check errors above.
See log: D:\projects\winforms2\artifacts\log\Debug\ToolsetRestore.binlog
```

</details>

**Comment by RussKie** [comment]
You should be able to build after https://github.com/dotnet/winforms/pull/4501.

**Comment by weltkante** [comment]
Still fails, notably it installs .NET 6 `6.0.100-preview.1.21077.13` but then tries to restore Arcade in `6.0.0-beta.21078.12` which it can't find. (Not sure if these version numbers are supposed to be related in any way.)

<details><summary>build log</summary>

```
D:\projects\winforms2 [main ≡]> .\build
Processing D:\projects\winforms2\global.json
cmake.exe already exists; replacing...
dotnet-api-docs_net5.0 was not found in .
Native tools bootstrap failed
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.1.21077.13/dotnet-sdk-6.0.100-preview.1.21077.13-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.1.21077.13/dotnet-sdk-6.0.100-preview.1.21077.13-win-x64.zip
dotnet-install: Adding to current process PATH: "D:\projects\winforms2\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install/windows#dependencies
dotnet-install: Installation finished
D:\projects\winforms2\.dotnet\sdk\6.0.100-preview.1.21077.13\MSBuild.dll /nologo -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,D:\projects\winforms2\.dotnet\sdk\6.0.100-preview.1.21077.13\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,D:\projects\winforms2\.dotnet\sdk\6.0.100-preview.1.21077.13\dotnet.dll -maxcpucount /m -verbosity:m /v:minimal /bl:D:\projects\winforms2\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:True /p:ContinuousIntegrationBuild=False /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=D:\projects\winforms2\artifacts\toolset\6.0.0-beta.21078.12.txt /t:__WriteToolsetLocation /warnaserror D:\projects\winforms2\artifacts\toolset\restore.proj
D:\projects\winforms2\artifacts\toolset\restore.proj : error : Das Paket "Microsoft.DotNet.Arcade.Sdk" der Version (= 6.0.0-beta.21078.12) wurde nicht gefunden.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 178 Version(en) gefunden in "dotnet-tools" [ Nächste Version: 5.0.0-beta.19558.13 ]
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 2 Version(en) gefunden in "dotnet5-transport" [ Nächste Version: 5.0.0-beta.20364.3 ]
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in nuget.org.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in richnav.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in dotnet5.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in dotnet6-transport.
D:\projects\winforms2\artifacts\toolset\restore.proj : error :   - 0 Version(en) gefunden in dotnet6.
D:\projects\winforms2\artifacts\toolset\restore.proj : error MSB4236: Das angegebene SDK "Microsoft.DotNet.Arcade.Sdk" wurde nicht gefunden.
Build failed with exit code 1. Check errors above.
See log: D:\projects\winforms2\artifacts\log\Debug\ToolsetRestore.binlog
```

</details>

**Comment by RussKie** [comment]
This is strange, it builds on the [CI](https://dev.azure.com/dnceng/public/_build/results?buildId=973484&view=results) and "works on my machine":

<details>
<summary>build.log</summary>

```
PS C:\Development\winforms> git clean -xdf
Removing .dotnet/
Removing .vs/
Removing artifacts/

PS C:\Development\winforms> .\Restore.cmd
Processing C:\Development\winforms\global.json
cmake.exe already exists; replacing...
dotnet-api-docs_net5.0 was not found in .
Native tools bootstrap failed
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.1.21077.13/dotnet-sdk-6.0.100-preview.1.21077.13-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.1.21077.13/dotnet-sdk-6.0.100-preview.1.21077.13-win-x64.zip
dotnet-install: Adding to current process PATH: "C:\Development\winforms\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install/windows#dependencies
dotnet-install: Installation finished
  Determining projects to restore...
  Restored C:\Users\igveliko\.nuget\packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21078.12\tools\Tools.proj (in 278 ms).
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Runtime/6.0.0-preview.2.21080.7/dotnet-runtime-6.0.0-preview.2.21080.7-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/6.0.0-preview.2.21080.7/dotnet-runtime-6.0.0-preview.2.21080.7-win-x64.zip
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install/windows#dependencies
dotnet-install: Installation finished
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Runtime/6.0.0-preview.2.21080.7/dotnet-runtime-6.0.0-preview.2.21080.7-win-x86.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/6.0.0-preview.2.21080.7/dotnet-runtime-6.0.0-preview.2.21080.7-win-x86.zip
dotnet-install: Adding to current process PATH: "C:\Development\winforms\.dotnet\x86\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install/windows#dependencies
dotnet-install: Installation finished
  Determining projects to restore...
  Determining projects to restore...
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj (in 844 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms.Primitives\src\System.Windows.Forms.Primitives.csproj (in 482 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\AccessibilityTests\AccessibilityTests.csproj (in 11 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\src\System.Windows.Forms.csproj (in 18 ms).
  Restored C:\Development\winforms\src\System.Drawing\src\System.Drawing.Facade.csproj (in 134 ms).
  Restored C:\Development\winforms\src\System.Drawing.Design\src\System.Drawing.Design.Facade.csproj (in 40 ms).
  Restored C:\Development\winforms\src\System.Design\src\System.Design.Facade.csproj (in 39 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiTestsHelper\MauiTestsHelper.csproj (in 1.62 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\System.Windows.Forms.IntegrationTests.Common\System.Windows.Forms.IntegrationTests.Common.csproj (in 1.62 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiButtonTests\MauiButtonTests.csproj (in 1.62 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiComboBoxTests\MauiComboBoxTests.csproj (in 1.62 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiListViewTests\MauiListViewTests.csproj (in 1.62 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiToolStripTests\MauiToolStripTests.csproj (in 1.62 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiMonthCalendarTests\MauiMonthCalendarTests.csproj (in 1.62 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiLabelTests\MauiLabelTests.csproj (in 1.62 sec).
  Restored C:\Development\winforms\pkg\Microsoft.Private.Winforms\Microsoft.Private.Winforms.csproj (in 24 ms).
  Restored C:\Development\winforms\src\Accessibility\ver\Accessibility-version.csproj (in 30 ms).
  Restored C:\Development\winforms\pkg\Microsoft.Dotnet.WinForms.ProjectTemplates\Microsoft.Dotnet.Winforms.ProjectTemplates.csproj (in 34 ms).
  Restored C:\Development\winforms\src\Microsoft.VisualBasic.Forms\src\Microsoft.VisualBasic.Forms.vbproj (in 48 ms).
  Restored C:\Development\winforms\src\Microsoft.VisualBasic\src\Microsoft.VisualBasic.Facade.csproj (in 78 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms.Design\src\System.Windows.Forms.Design.csproj (in 34 ms).
  Restored C:\Development\winforms\src\Accessibility\src\Accessibility.ilproj (in 100 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms.Design.Editors\src\System.Windows.Forms.Design.Editors.Facade3x.csproj (in 75 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiPropertyGridViewTests\MauiPropertyGridViewTests.csproj (in 72 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\MauiTests\MauiPrintPreviewDialogTests\MauiPrintPreviewDialogTests.csproj (in 67 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\WinformsControlsTest\WinformsControlsTest.csproj (in 205 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms.SourceGenerators\src\System.Windows.Forms.SourceGenerators.csproj (in 908 ms).
  Restored C:\Development\winforms\src\Microsoft.VisualBasic\tests\IntegrationTests\VisualBasicRuntimeTest\VisualBasicRuntimeTest.csproj (in 9 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\TestUtilities\System.Windows.Forms.TestUtilities.csproj (in 3.18 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms.Primitives\tests\TestUtilities\System.Windows.Forms.Primitives.TestUtilities.csproj (in 2.32 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms.SourceGenerators\tests\UnitTests\System.Windows.Forms.SourceGenerators.Tests.csproj (in 1.4 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\System.Windows.Forms.IntegrationTests\System.Windows.Forms.IntegrationTests.csproj (in 1.47 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms.Primitives\tests\TestUtilities.Tests\System.Windows.Forms.Primitives.TestUtilities.Tests.csproj (in 1.58 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\System.Windows.Forms.Interop.Tests.csproj (in 3.25 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\System.Windows.Forms.Maui.IntegrationTests\System.Windows.Forms.Maui.IntegrationTests.csproj (in 1.52 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms.Design\tests\UnitTests\System.Windows.Forms.Design.Tests.csproj (in 1.56 sec).
  Restored C:\Development\winforms\src\Microsoft.VisualBasic\tests\UnitTests\Microsoft.VisualBasic.Tests.csproj (in 1.32 sec).
  Restored C:\Development\winforms\src\System.Windows.Forms.Primitives\tests\UnitTests\System.Windows.Forms.Primitives.Tests.csproj (in 1.4 sec).
  Restored C:\Development\winforms\src\Microsoft.VisualBasic\tests\IntegrationTests\Microsoft.VisualBasic.IntegrationTests\Microsoft.VisualBasic.IntegrationTests.csproj (in 437 ms).
  Restored C:\Development\winforms\src\System.Windows.Forms\tests\UnitTests\System.Windows.Forms.Tests.csproj (in 3.26 sec).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:36.83
PS C:\Development\winforms>
```

</details>

Try maybe deleting C:\Users\\<user\>\\.nuget\packages\microsoft.dotnet.arcade.sdk?

**Comment by weltkante** [comment]
Nope, the 6.0 package is not in there, deleting it doesn't help anything either, opened dotnet/arcade#6887

**Comment by weltkante** [comment]
I've [experimented a bit](https://github.com/dotnet/winforms/compare/main...weltkante:issue3595) with the idea and noticed the (in hindsight obvious) fact that it is actual normal operation to destroy handles while the WndProc is still running (e.g. every form destroys itself in response to `WM_CLOSE`, or a button click may close its form, which in turn destroys the button)

So the operation to destroy (or recreate) the handle while your WndProc is on stack is normal for some messages, but fatal for other messages. That doesn't make the suggested approach useless, but it needs to be selectively applied by [adding protection to specific handlers](https://github.com/dotnet/winforms/compare/main...weltkante:issue3595#diff-a279bfceaf830444fc4318432b3387481fe38b4b5fdde61a216bdd3bf8a944feR3621) like `CBN_DROPDOWN`.

I've [tried to implement](https://github.com/dotnet/winforms/compare/main...weltkante:issue3595#diff-07a0a87cedab0d76c974ce8b105912a1b986c87116c7ee0ac73d6d5d65e4b48aR5098) the idea but it doesn't fit nicely into NativeWindow because it maintains a strong association with its current handle (WndProc callbacks are bound to individual NativeWindow instances and cannot be transferred or changed at arbitrary times), and in turn the owning Control maintains a strong association with its NativeWindow (by making the variable readonly and the ControlNativeWindow referring back to the Control instance there is a symmetric one-to-one association).

Solving this gracefully would probably require redesigning the Control/NativeWindow infrastructure to introduce a detached "zombie" state for NativeWindow and for either the Control or NativeWindow being able to support multiple handles existing at the same time. I'm not very keen on touching the NativeWindow in this way, its logic looks very fragile.

If you just want to avoid the AccessViolations (being indeterministic, showing up only sometimes, and having potential for memory corruption) then you could keep the [protection infrastructure](https://github.com/dotnet/winforms/compare/main...weltkante:issue3595#diff-55e243b794f71728a69da9d0e151c86ea9d5204f34789f54f7f2bb4f8b31b7d3R908) and [throw early](https://github.com/dotnet/winforms/compare/main...weltkante:issue3595#diff-07a0a87cedab0d76c974ce8b105912a1b986c87116c7ee0ac73d6d5d65e4b48aR9572) in the RecreateHandle/DestroyHandle code instead.

I hope this makes sense? Its hard to communicate succinctly which approaches I tried and why they didn't work out. If you have suggestions or questions feel free to comment; its not that I think its impossible to do, its just that I don't like the risk involved in introducing an additional state in the NativeWindow lifecycle, and I think test driven development (coding an implementation and see if anything breaks) is very likely to miss lots of edge cases.
